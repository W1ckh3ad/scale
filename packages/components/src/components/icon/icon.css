.icon path {
  transition: all var(--scl-motion-duration-fast)
    var(--scl-motion-easing-default);
}
