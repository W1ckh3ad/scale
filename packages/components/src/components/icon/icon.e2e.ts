import { newE2EPage } from '@stencil/core/testing';

describe('scale-icon', () => {
  it('renders', async () => {
    const page = await newE2EPage();
    await page.setContent('<scale-icon path="m 10 10">Label</scale-icon>');
    const element = await page.find('scale-icon');
    expect(element).toHaveClass('hydrated');
  });
});
