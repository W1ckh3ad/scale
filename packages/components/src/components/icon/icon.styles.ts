import { JssStyle } from 'jss';

export const styles: JssStyle = {
  icon: {
    '& path': {
      transition: 'all .2s ease-in-out',
    },
  },
};
