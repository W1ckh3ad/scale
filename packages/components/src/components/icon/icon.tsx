import { Component, Prop, h, Host } from '@stencil/core';
import classNames from 'classnames';

@Component({
  tag: 'scale-icon',
  styleUrl: './icon.css',
})
export class Icon {
  /**
   * A name that will be used to reference an SVG object defined elsewhere,
   * via `<use xlink:href="">`. `icon-` will be prepended to the name, so if
   * you pass `circle`, it will look for for the `icon-circle` id
   * e.g. `<use xlink:href="#icon-circle">`.
   * If there is no element in the document with the id by the name provided,
   * this component will render empty.
   */
  @Prop() name?: string;
  /**
   * A path shape to be used in the `d` attribute of a path element.
   */
  @Prop() path?: string;
  /**
   * Will be used for both `width` and `height`, all icons are square.
   * Keep in mind the `viewBox` attribute is set to "0 0 24 24".
   */
  @Prop() size?: number = 24;
  /** The SVG `fill` attribute */
  @Prop() fill?: string = 'var(--icon-color, currentColor)';
  /** The SVG `stroke` attribute */
  @Prop() stroke?: string = 'transparent';
  /** (optional) If `true` the icon can receive focus */
  @Prop() focusable?: boolean = false;
  /** (optional) If `true` the svg element will get aria-hidden="true" */
  @Prop() decorative?: boolean = false;
  /** (optional) When using the icon as standalone, make it meaningful for accessibility */
  @Prop() accessibilityTitle?: string;

  render() {
    const pathAttributes = {
      fill: this.fill,
      stroke: this.stroke,
    };
    const ariaHidden = this.decorative ? { 'aria-hidden': 'true' } : {};

    return (
      <Host style={{ display: 'inline-flex' }}>
        <svg
          xmlns="http://www.w3.org/2000/svg"
          {...(this.focusable ? { tabindex: 0 } : {})}
          class={this.getCssClassMap()}
          width={this.size}
          height={this.size}
          viewBox="0 0 24 24"
          role="img"
          {...ariaHidden}
        >
          {this.accessibilityTitle && <title>{this.accessibilityTitle}</title>}
          {this.path ? (
            <path d={this.path} {...pathAttributes} />
          ) : (
            <use xlinkHref={`#icon-${this.name}`} {...pathAttributes} />
          )}
        </svg>
      </Host>
    );
  }

  getCssClassMap() {
    return classNames('icon');
  }
}
