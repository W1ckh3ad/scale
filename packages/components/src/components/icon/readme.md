# scale-icon



<!-- Auto Generated Below -->


## Properties

| Property      | Attribute      | Description                                                                                                                                                                                                                                                                                                                                                    | Type      | Default                             |
| ------------- | -------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------- | ----------------------------------- |
| `ariaLabel`   | `aria-label`   | (optional) When using the icon by itself, add a label to improve accessibility                                                                                                                                                                                                                                                                                 | `string`  | `undefined`                         |
| `customClass` | `custom-class` | (optional) Tag custom class                                                                                                                                                                                                                                                                                                                                    | `string`  | `''`                                |
| `fill`        | `fill`         | The SVG `fill` attribute                                                                                                                                                                                                                                                                                                                                       | `string`  | `'var(--icon-color, currentColor)'` |
| `focusable`   | `focusable`    | (optional) If `true` the icon can receive focus                                                                                                                                                                                                                                                                                                                | `boolean` | `false`                             |
| `name`        | `name`         | A name that will be used to reference an SVG object defined elsewhere, via `<use xlink:href="">`. `icon-` will be prepended to the name, so if you pass `circle`, it will look for for the `icon-circle` id e.g. `<use xlink:href="#icon-circle">`. If there is no element in the document with the id by the name provided, this component will render empty. | `string`  | `undefined`                         |
| `path`        | `path`         | A path shape to be used in the `d` attribute of a path element.                                                                                                                                                                                                                                                                                                | `string`  | `undefined`                         |
| `size`        | `size`         | Will be used for both `width` and `height`, all icons are square. Keep in mind the `viewBox` attribute is set to "0 0 24 24".                                                                                                                                                                                                                                  | `number`  | `24`                                |
| `stroke`      | `stroke`       | The SVG `stroke` attribute                                                                                                                                                                                                                                                                                                                                     | `string`  | `'transparent'`                     |
| `styles`      | `styles`       | (optional) Injected jss styles                                                                                                                                                                                                                                                                                                                                 | `any`     | `undefined`                         |


## Dependencies

### Used by

 - [app-navigation-main-mobile](../app-navigation-main-mobile)
 - [scale-app-header](../app-header)
 - [scale-button](../button)
 - [scale-input](../input)
 - [scale-link](../link)
 - [scale-list-item](../list/listItem)
 - [scale-progress-bar](../progress-bar)
 - [scale-tag](../tag)

### Graph
```mermaid
graph TD;
  app-navigation-main-mobile --> scale-icon
  scale-app-header --> scale-icon
  scale-button --> scale-icon
  scale-input --> scale-icon
  scale-link --> scale-icon
  scale-list-item --> scale-icon
  scale-progress-bar --> scale-icon
  scale-tag --> scale-icon
  style scale-icon fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
