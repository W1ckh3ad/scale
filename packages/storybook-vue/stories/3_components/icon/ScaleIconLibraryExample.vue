<template>
  <div>
    <scale-icon-action-favorite
      :color="color"
      :size="size"
      :selected="selected"
      :decorative="decorative"
      :accessibility-title="accessibilityTitle"
    />
  </div>
</template>

<script>
export default {
  props: {
    color: { type: String, default: 'currentColor' },
    size: { type: Number, default: 24 },
    selected: { type: Boolean, default: false },
    decorative: { type: Boolean, default: false },
    accessibilityTitle: { type: String },
  }
}
</script>