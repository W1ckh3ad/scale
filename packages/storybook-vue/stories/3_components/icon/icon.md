# Icon

With icons our customers can easily find their way around our product or page and grasp content and functions quickly.

## General

Our icons fit harmoniously into the design of our websites and apps. They illustrate abstract and complex topics. With our own icon style, we subtly convey the values of the Telekom brand.

## Basic Types

### Static

A static icon has no interactive functionality.

![Image Name](/assets/3_components/icon/image-20200810101050303.png)

### Interactive

An interactive icon can have an assigned function.

![Image Name](/assets/3_components/icon/image-20200811094920205.png)

## Elements

### Description

An icon must always have a description so that it can be read out by screen readers. You may also display a tooltip.

## Sizes

### Large

48px

![Image Name](/assets/3_components/icon/image-20200810101050303.png)

### Medium

24px

![Image Name](/assets/3_components/icon/image-20200810101057046.png)

### Small

16px

![Image Name](/assets/3_components/icon/image-20200810101102823.png)

## Related Components
