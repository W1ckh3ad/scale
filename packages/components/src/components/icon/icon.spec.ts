import { newSpecPage } from '@stencil/core/testing';
import { Icon } from './icon';
import { styles } from './icon.styles';
import jss from 'jss';

describe('Icon', () => {
  let element;
  beforeEach(async () => {
    element = new Icon();
    element.stylesheet = jss.createStyleSheet(styles as any);
  });

  it('should match snapshot', async () => {
    const page = await newSpecPage({
      components: [Icon],
      html: `<scale-icon path="m 10 10">Label</scale-icon>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('should handle css classes', () => {
    element.customClass = 'custom';
    expect(element.getCssClassMap()).toContain('custom');

    element.name = 'arrow-left';
    expect(element.getCssClassMap()).toContain('arrow-left');
  });
});
