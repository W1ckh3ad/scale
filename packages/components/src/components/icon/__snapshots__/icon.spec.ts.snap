// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon should match snapshot 1`] = `
<scale-icon path="m 10 10" style="display: inline-flex;">
  Label
  <style>
    .icon-0-0-2 {}
.icon-0-0-2 path {
  transition: all .2s ease-in-out;
}
  </style>
  <svg class="icon-0-0-2" height="24" viewBox="0 0 24 24" width="24" xmlns="http://www.w3.org/2000/svg">
    <path d="m 10 10" fill="var(--icon-color, currentColor)" stroke="transparent"></path>
  </svg>
</scale-icon>
`;
